pkg/controllers/leaderworkerset_controller.go

```diff
@@ -206,107 +206,103 @@ func SetupIndexes(indexer client.FieldIndexer) error {
 //     together with scaling down.
 //
 // At rest, Partition should always be zero.
-func (r *LeaderWorkerSetReconciler) rollingUpdatePartition(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (int32, error) {
+//
+// For Replicas:
+//   - When rolling update, Replicas is equal to (spec.Replicas+maxSurge)
+//   - Otherwise, Replicas is equal to spec.Replicas
+//   - One exception here is when unready replicas of leaderWorkerSet is equal to MaxSurge,
+//     we should reclaim the extra replicas gradually to accommodate for the new replicas.
+func (r *LeaderWorkerSetReconciler) rollingUpdateParameters(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (partition int32, replicas int32, err error) {
+	lwsReplicas := *lws.Spec.Replicas
+
 	sts := &appsv1.StatefulSet{}
-	err := r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
+	err = r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
 	if err != nil {
-		// If sts not created yet, all partitions should be updated.
+		// Case 1:
+		// If sts not created yet, all partitions should be updated,
+		// replicas should not change.
 		if apierrors.IsNotFound(err) {
-			return 0, nil
+			return 0, lwsReplicas, nil
 		}
-		return 0, err
+		return 0, 0, err
 	}
 
-	partition := *sts.Spec.UpdateStrategy.RollingUpdate.Partition
-	replicas := *lws.Spec.Replicas
-	rollingStep, err := intstr.GetValueFromIntOrPercent(&lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxUnavailable, int(replicas), false)
+	stsReplicas := *sts.Spec.Replicas
+	maxSurge, err := intstr.GetValueFromIntOrPercent(&lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxSurge, int(lwsReplicas), false)
 	if err != nil {
-		return 0, err
+		return 0, 0, err
+	}
+	// No need to burst more than the replicas.
+	if maxSurge > int(lwsReplicas) {
+		maxSurge = int(lwsReplicas)
 	}
+	burstReplicas := lwsReplicas + int32(maxSurge)
+
+	var continuousReadyReplicas, lwsUnreadyReplicas int32
+
+	defer func() {
+		// Reclaim the bursted replicas gradually.
+		if lwsUnreadyReplicas <= int32(maxSurge) {
```

Comment:
I find it hard to reason about when using a defer function and named return variables, I suggest to directly put the value in the return statements below and remove the named return variable and defer func. 

Also, I am not sure the calculation here is correct, shouldn't it always be the following (without the if statement above):

```replicas = lwsReplicas + min(utils.NonZeroValue(lwsUnreadyReplicas-maxUnavailable), maxSurge)```